ACPI changes

sameo opened the pull request:
This is a PR created for reviewing convenience. It contains our first batch of ACPI patches that we want to send for upstream review. It includes all of our generic ACPI changes and refactoring, together with our hardware-reduced ACPI implementation.

Commit:
hw: i386: Decouple the ACPI build from the PC machine type

ACPI tables are platform and machine type and even architecture
agnostic, and as such we want to provide an internal ACPI API that
only depends on platform agnostic information.

For the x86 architecture, in order to build ACPI tables independently
from the PC or Q35 machine types, we are moving a few MachineState
structure fields into a machine type agnostic structure called
AcpiConfiguration. The structure fields we move are:

   HotplugHandler *acpi_dev
   AcpiNVDIMMState acpi_nvdimm_state;
   FWCfgState *fw_cfg
   ram_addr_t below_4g_mem_size, above_4g_mem_size
   bool apic_xrupt_override
   unsigned apic_id_limit
   uint64_t numa_nodes
   uint64_t numa_mem

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Cc: Shannon Zhao <zhaoshenglong@huawei.com>
Cc: Peter Maydell <peter.maydell@linaro.org>
Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Richard Henderson <rth@twiddle.net>
Cc: Eduardo Habkost <ehabkost@redhat.com>
Cc: Marcel Apfelbaum <marcel.apfelbaum@gmail.com>
Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: acpi: Initial hardware-reduced support

We build a minimal set of ACPI hardware-reduced tables: XSDT,
FADT, MADT and a DSDT pointed by a RSDP.
The DSDT only contains one PCI host bridge for now.

This API will be consumed by new x86 machine type but also potentially
by the ARM virt one.

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: acpi: Define ACPI tables builder interface

In order to decouple ACPI APIs from specific machine types, we are
creating an ACPI builder interface that each ACPI platform can choose to
implement.
This way, a new machine type can re-use the high level ACPI APIs and
define some custom table build methods, without having to duplicate most
of the existing implementation only to add small variations to it.

Cc: Eduardo Habkost <ehabkost@redhat.com>
Cc: Marcel Apfelbaum <marcel.apfelbaum@gmail.com>
Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: arm: Switch to the AML build RSDP building routine

We make the ARM virt ACPI code use the now shared build_rsdp() API from
aml-build.c. By doing so we fix a bug where the ARM implementation was
missing adding both the legacy and extended checksums, which was
building an invalid RSDP table.

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Cc: Shannon Zhao <zhaoshenglong@huawei.com>
Cc:Peter Maydell <peter.maydell@linaro.org>
Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

yangzhon committed:
hw: i386: Refactor PCI host getter

Make it more flexible by having it parsing a PCI host paths array
instead of open coding those paths deep down into the code logic itself.
This will be needed for PCI machine types that are neither emulatiing the
ich9 nor the i440fx chipsets.

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Richard Henderson <rth@twiddle.net>
Cc: Eduardo Habkost <ehabkost@redhat.com>
Cc: Marcel Apfelbaum <marcel.apfelbaum@gmail.com>
Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: acpi: Generalize AML build routines

Most of the AML build routines under acpi-build are not even
architecture specific. They can be moved to the more generic hw/acpi
folder where they could be shared across machine types and
architectures.

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Richard Henderson <rth@twiddle.net>
Cc: Eduardo Habkost <ehabkost@redhat.com>
Cc: Marcel Apfelbaum <marcel.apfelbaum@gmail.com>
Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: acpi: Factorize _OSC AML across architectures

The _OSC AML table is almost identical between the i386 Q35 and arm virt
machine types. We can make it slightly more generic and share it across
all PCIe architectures.

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Cc: Shannon Zhao <zhaoshenglong@huawei.com>
Cc: Peter Maydell <peter.maydell@linaro.org>
Cc: Marcel Apfelbaum <marcel.apfelbaum@gmail.com>
Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Richard Henderson <rth@twiddle.net>
Cc: Eduardo Habkost <ehabkost@redhat.com>
Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: acpi: Export and generalize the PCI host AML API

The AML build routines for the PCI host bridge and the corresponding
DSDT addition are neither x86 nor PC machine type specific.
We can move them to the architecture agnostic hw/acpi folder, and by
carrying all the needed information through a new AcpiPciBus structure,
we can make them PC machine type independent.

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Cc: Marcel Apfelbaum <marcel.apfelbaum@gmail.com>
Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Richard Henderson <rth@twiddle.net>
Cc: Eduardo Habkost <ehabkost@redhat.com>
Signed-off-by: Yang Zhong <yang.zhong@intel.com>
Signed-off-by: Rob Bradford <robert.bradford@intel.com>

yangzhon committed:
hw: acpi: Export the MCFG getter

The ACPI MCFG getter is not x86 specific and could be called from
anywhere within generic ACPI API, so let's export it.

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Richard Henderson <rth@twiddle.net>
Cc: Eduardo Habkost <ehabkost@redhat.com>
Cc: Marcel Apfelbaum <marcel.apfelbaum@gmail.com>
Signed-off-by: Yang Zhong <yang.zhong@intel.com>

Commit:
hw: acpi: reduced: Add shutdown support

Hardware-reduced ACPI uses SLEEP_CONTROL_REG to enter S5 sleep state.

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

Commit:
hw: acpi: reduced: Generic Event Device support

The ACPI Generic Event Device (GED) is a hardware-reduced specific
device that handles all platform events, including the hotplug ones.
This patch generate the AML code that defines GEDs.
Platforms need to specify their own GedEvent array to describe what kind
of events they want to support through GED. The build_ged_aml routine
takes a GedEvent array that maps a specific GED event to an IRQ number.
Then we use that array to build both the _CRS and the _EVT section
of the GED device.

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: acpi: reduced: Add reboot support

We only need to expose it through FADT.

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

yangzhon committed:
hw: acpi: reduced: Add memory hotplug support

We add the memory hotplug AML code to the hardware-reduced DSDT.
The memory hotplug event is handled through the GED device.

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Signed-off-by: Yang Zhong <yang.zhong@intel.com>

Commit:
hw: acpi: Do not create hotplug method when handler is not defined

CPU and memory ACPI hotplug are not necessarily handled through SCI
events. For example, with Hardware-reduced ACPI, the GED device will
manage ACPI hotplug entirely.
As a consequence, we make the CPU and memory specific events AML
generation optional. The code will only be added when the method name is
not NULL.

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: acpi: reduced: Add NFIT support

If the platform is NVDIMM enabled, we add the NFIT table to the ACPI
build.

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>

yangzhon committed:
hw: acpi: Export the SRAT AML build API

The SRAT ACPI table is not x86 specific and will be needed for the
Hardware-reduced ACPI implementation. So we should export it through the
architecture independent hw/acpi folder.

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Richard Henderson <rth@twiddle.net>
Cc: Eduardo Habkost <ehabkost@redhat.com>
Cc: Marcel Apfelbaum <marcel.apfelbaum@gmail.com>
Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: acpi: reduced: Add SRAT table

If the platform is a NUMA enabled, we add the SRAT table to the ACPI
build. It is up to the calling platform to define its own SRAT build
method or use the aml-build.c one.

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Signed-off-by: Yang Zhong <yang.zhong@intel.com>

yangzhon committed:
hw: acpi: Fix memory hotplug AML generation error

When using the generated memory hotplug AML, the iasl
compiler would give the following error:

dsdt.dsl 266: Return (MOST (_UID, Arg0, Arg1, Arg2))
Error 6080 - Called method returns no value ^

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Signed-off-by: Yang Zhong <yang.zhong@intel.com>

Commit:
hw: acpi: reduced: Add PCI hotplug support

In order to support PCI hotplug through a hardware-reduced GED, we need
to modify the GED device definition in the DSDT table, so that a PCI hotplug
related interrupt will trigger a new PCI scan.
We also need to modify the DSDT PCI bus definition in order to make sure
a PCI scan of all available slots can be performed when an interrupt
comes in.

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>
Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

Commit:
hw: acpi: Export the PCI hotplug API

The ACPI hotplug support for PCI devices APIs are not x86 or even
machine type specific. In order for future machine types to be able to
re-use that code, we export it through the architecture agnostic
hw/acpi folder.

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Richard Henderson <rth@twiddle.net>
Cc: Eduardo Habkost <ehabkost@redhat.com>
Cc: Marcel Apfelbaum <marcel.apfelbaum@gmail.com>
Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>
Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

Commit:
hw: acpi: Retrieve the PCI bus from AcpiPciHpState

Instead of using the machine type specific method find_i440fx() to
retrieve the PCI bus, this commit aims to rely on the fact that the
PCI bus is known by the structure AcpiPciHpState.

When the structure is initialized through acpi_pcihp_init() call,
it saves the PCI bus, which means there is no need to invoke a
special function later on.

Based on the fact that find_i440fx() was only used there, this
patch also removes the function find_i440fx() itself from the
entire codebase.

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Cc: Marcel Apfelbaum <marcel.apfelbaum@gmail.com>
Cc: Paolo Bonzini <pbonzini@redhat.com>
Signed-off-by: Sebastien Boeuf <sebastien.boeuf@intel.com>
Signed-off-by: Jing Liu <jing2.liu@linux.intel.com>

Commit:
hw: i386: Make the hotpluggable memory size property more generic

This property is currently defined under i386/pc while it only describes
a region size that's eventually fetched from the AML ACPI code.

We can make it more generic and shareable across machine types by moving
it to memory-device.h instead.

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Cc: Marcel Apfelbaum <marcel.apfelbaum@gmail.com>
Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Richard Henderson <rth@twiddle.net>
Cc: Eduardo Habkost <ehabkost@redhat.com>
Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: acpi: Export ACPI build alignment API

This is going to be needed by the Hardware-reduced ACPI routines.

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Cc: Marcel Apfelbaum <marcel.apfelbaum@gmail.com>
Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Richard Henderson <rth@twiddle.net>
Cc: Eduardo Habkost <ehabkost@redhat.com>
Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

Commit:
hw: i386: Convert PC machine type to firmware build methods

All PC machine type derivatives will use the same ACPI table build
methods. But with that change in place, any new x86 machine type will be
able to re-use the acpi-build API and customize part of it by defining
its own ACPI table build methods.

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Cc: Marcel Apfelbaum <marcel.apfelbaum@gmail.com>
Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Richard Henderson <rth@twiddle.net>
Cc: Eduardo Habkost <ehabkost@redhat.com>
Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

yangzhon committed:
hw: acpi: reduced: Add MCFG support

For PCIe based platform, we need to add an MCFG table to the
hardware-reduced DSDT.

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Signed-off-by: Yang Zhong <yang.zhong@intel.com>
Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

sameo commented:
cc @bonzini, in case you're interested. We'll send a version of this PR to the qemu mailing list very shortly.

liujing2 commented on hw/acpi/reduced.c:
```diff
@@ -44,8 +46,30 @@
+    };
+
+    build_cpus_aml(dsdt, ms, opts, conf->cpu_hotplug_io_base,
+                   "\\_SB.PCI1", GED_DEVICE);
```
@sameo  Does \\_SB.PCI0 work?

liujing2 replied:
@sameo I found later this would be changed to \\_SB, but what I concern is if PCI1 will be a discussion issue in this patch.

sameo replied:
@liujing2 This was a mistake from my part. This now fixed, we never use `PCI1`.

rbradford commented on include/hw/acpi/reduced.h:
```diff
@@ -0,0 +1,6 @@
+#ifndef HW_ACPI_REDUCED_H
```
Does this file need a copyright header?

sameo replied:
Fixed.

rbradford commented on hw/acpi/reduced.c:
```diff
@@ -0,0 +1,245 @@
+
+    acpi_get_pci_holes(&pci_hole, &pci_hole64);
+    table_offsets = g_array_new(false, true /* clear */,
+                                        sizeof(uint32_t));
```
Is this consistent with the coding style?

liujing2 replied:
Is it good to use ./scripts/checkpatch.pl to check if our patches match the coding style? I also see many lines more of 72 characters.

sameo replied:
All patches are `checkpatch` clean now, thanks for the reminder.

rbradford commented on include/hw/boards.h:
```diff
@@ -5,6 +5,7 @@
 
```
Typo of generic in the commit message for 3d240bcd9dbb781fafd5c2d2093a56e5e7dee44f

sameo replied:
Fixed.

rbradford commented on hw/acpi/aml-build.c:
```diff
@@ -1869,51 +1869,51 @@ Aml *build_crs(PCIHostState *host, CrsRangeSet *range_set)
     return crs;
 }
 
```
Typo of slightly in commit message 861742a39d538b9919967a9bedee8b28a4475e7d

sameo replied:
Fixed.

rbradford commented on hw/acpi/aml-build.c:
```diff
@@ -2455,6 +2456,134 @@ void build_srat_memory(AcpiSratMemoryAffinity *numamem, uint64_t base,
     numamem->range_length = cpu_to_le64(len);
 }
 
```
I think the commit message for 88762b272b20f5c331263e87ce5d4c4e4d061174 is wrong, i think it should be architecture *independent*?

sameo replied:
Yes, fixed. Thanks.

rbradford commented on hw/acpi/aml-build.c:
```diff
@@ -2310,6 +2311,199 @@ Aml *build_pci_host_bridge(Aml *table, AcpiPciBus *pci_host)
+}
+
+void build_append_pci_bus_devices(Aml *parent_scope, PCIBus *bus,
+                                         bool pcihp_bridge_en)
```
Something looks funky with the whitespace here.

sameo replied:
Fixed as well.

rbradford reviewed (commented):
Please double-check with the QEMU guidelines whether you need to include your Signed-off-by on the patches from other people.

bonzini commented:
Why are the hardware-reduced tables included, since they are only used by the virt machine type?

mcastelino commented on include/hw/acpi/acpi.h:
```diff
@@ -186,6 +188,40 @@ extern int acpi_enabled;
+    /* Machine state settings */
+    FWCfgState *fw_cfg;
+    HotplugHandler *acpi_dev;
+    ram_addr_t below_4g_mem_size, above_4g_mem_size;;
```
Are these fields specific to ACPI
```
    ram_addr_t below_4g_mem_size, above_4g_mem_size;;
    uint64_t numa_nodes;
    uint64_t *node_mem;
    bool apic_xrupt_override;
    unsigned apic_id_limit;
```
or just moved here for convenience.

mcastelino replied:
Also it would be helpful for the commit message itself to be expanded to explain what got moved around and why.

sameo replied:
> Are these fields specific to ACPI
> 
> ```
>     ram_addr_t below_4g_mem_size, above_4g_mem_size;;
>     uint64_t numa_nodes;
>     uint64_t *node_mem;
>     bool apic_xrupt_override;
>     unsigned apic_id_limit;
> ```
> or just moved here for convenience.

They're used for building ACPI tables, so not ACPI specific fields. I'll fix the structure to make it clearer.

sameo replied:
> Also it would be helpful for the commit message itself to be expanded to explain what got moved around and why.

Done.

mcastelino commented on hw/i386/acpi-build.c:
```diff
@@ -2803,72 +2795,61 @@ static void acpi_build_reset(void *build_opaque)
     .minimum_version_id = 1,
     .fields = (VMStateField[]) {
-        VMSTATE_UINT8(patched, AcpiBuildState),
+        VMSTATE_BOOL(patched, AcpiBuildState),
```
Why was this change needed?

sameo replied:
Both the arm and i386 machine type had a `AcpiBuildState` and the only difference was the `patched` type. i386 was using a uint8 while arm was using a boolean. When factorizing those, it made more sense to keep it as boolean.

mcastelino commented on hw/acpi/aml-build.c:
```diff
@@ -1651,6 +1651,69 @@ build_xsdt(GArray *table_data, BIOSLinker *linker, GArray *table_offsets,
                  (void *)xsdt, "XSDT", xsdt_len, 1, oem_id, oem_table_id);
 }
 
+/* Legacy RSDP pointing at an RSDT. This is deprecated */
```
Would be good to expand the commit message to explain why the code motion was needed. Or maybe split the code motion into a separate commit.

sameo replied:
I split the patch in two, thanks for the suggestion.

sameo replied:
I split the patch in two, thanks for the suggestion.

mcastelino commented on hw/acpi/aml-build.c:
```diff
@@ -1651,6 +1651,69 @@ build_xsdt(GArray *table_data, BIOSLinker *linker, GArray *table_offsets,
+}
+
+/* RSDP pointing at an XSDT */
+GArray *build_rsdp(GArray *rsdp_table, BIOSLinker *linker, unsigned xsdt_tbl_offset)
```
Was build_rsdp modified in any way or just moved around?

mcastelino commented on hw/arm/virt-acpi-build.c:
```diff
@@ -35,6 +35,7 @@
 #include "target/arm/cpu.h"
 #include "hw/acpi/acpi-defs.h"
 #include "hw/acpi/acpi.h"
+#include "hw/acpi/aml-build.h"
```
So you are replacing the implementation here with the new implementation that was added earlier. Maybe expand the commit message to indicate the same.

sameo replied:
Done.

mcastelino commented on hw/acpi/aml-build.c:
```diff
@@ -1869,51 +1869,51 @@ Aml *build_crs(PCIHostState *host, CrsRangeSet *range_set)
 }
 
-Aml *build_osc_method(void)
+Aml *build_osc_method(uint32_t value)
```
Could use a comment/header on what value was being used for.
I assume this is the `_OSC Capabilities Buffer Control Field` mask.

sameo replied:
Done

mcastelino commented on hw/i386/acpi-build.c:
```diff
@@ -1335,7 +1335,9 @@ build_dsdt(GArray *table_data, BIOSLinker *linker,
-        aml_append(dev, build_osc_method());
+        aml_append(dev, aml_name_decl("SUPP", aml_int(0)));
+        aml_append(dev, aml_name_decl("CTRL", aml_int(0)));
+        aml_append(dev, build_osc_method(0x1F));
```
I assume this is the `_OSC Control Field` mask.
It may be good to not make it a manifest constant. Instead an ORing of defined constants.

I assume this indicates the following are supported
```
0 PCI Express Native Hot Plug control
1 SHPC Native Hot Plug control
2 PCI Express Native Power Management Events control
3 PCI Express Advanced Error Reporting control
4 PCI Express Capability Structure control
```
and
```
5 Latency Tolerance Reporting control
6 are ignored
```

/cc @sboeuf

sameo replied:
Done

mcastelino commented on hw/acpi/aml-build.c:
```diff
@@ -1869,51 +1869,51 @@ Aml *build_crs(PCIHostState *host, CrsRangeSet *range_set)
-
-    aml_append(if_ctx, aml_store(aml_name("CDW3"), a_ctrl));
-
-    /*
```
We lost these comments which provided some context.

mcastelino commented on hw/acpi/aml-build.c:
```diff
@@ -29,6 +29,25 @@
+#define IO_0_LEN                           0xcf8
+#define VGA_MEM_LEN                        0x20000
+
+static const char *pci_hosts[] = {
```
Oh so we moved it around again. This confused me for a bit.

mcastelino commented on hw/acpi/aml-build.c:
```diff
@@ -2095,6 +2159,150 @@ Aml *build_prt(bool is_pci0_prt)
     return method;
 }
 
+Aml *build_pci_host_bridge(Aml *table, AcpiPciBus *pci_host)
```
We seem a little inconsistent with how we refer to AcpiPciBus, we call it  pci_host in some cases and differently in other places. Let me see if I can find that.

sameo replied:
We call it acpi_pci_host in `reduced.c:acpi_reduced_build()`, because there already is a `pci_host` pointer.

mcastelino commented on hw/acpi/cpu.c:
```diff
@@ -569,9 +569,11 @@ void build_cpus_aml(Aml *table, MachineState *machine, CPUHotplugFeatures opts,
-    aml_append(table, method);
+    if (event_handler_method) {
+        method = aml_method(event_handler_method, 0, AML_NOTSERIALIZED);
+        aml_append(method, aml_call0("\\_SB.CPUS." CPU_SCAN_METHOD));
```
Wish we did something like `MEMORY_DEVICES_CONTAINER` out here too. maybe `CPU_DEVICES_CONTAINER`

mcastelino commented on hw/acpi/aml-build.c:
```diff
@@ -2455,6 +2456,134 @@ void build_srat_memory(AcpiSratMemoryAffinity *numamem, uint64_t base,
+    MachineClass *mc = MACHINE_GET_CLASS(machine);
+    const CPUArchIdList *apic_ids = mc->possible_cpu_arch_ids(machine);
+    ram_addr_t hotplugabble_address_space_size =
+        object_property_get_int(OBJECT(machine), PC_MACHINE_DEVMEM_REGION_SIZE,
```
PC_MACHINE_DEVMEM_REGION_SIZE does not seem to be architecture independent.

sameo replied:
Fixed as well.

mcastelino commented on hw/acpi/reduced.c:
```diff
@@ -30,6 +30,8 @@
 #include "hw/acpi/acpi.h"
 #include "hw/acpi/aml-build.h"
 #include "hw/acpi/bios-linker-loader.h"
+#include "hw/acpi/cpu.h"
```
This commit could have been part of different patch set.

mcastelino commented on hw/acpi/reduced.c:
```diff
@@ -249,3 +275,136 @@ void acpi_reduced_setup(MachineState *machine, AcpiConfiguration *conf)
     acpi_build_tables_cleanup(&tables, false);
 }
+
+#define CPU_SCAN_METHOD   "CSCN"
```
@sboeuf is this what you are calling new vs legacy?

sboeuf replied:
Yes, this way we should not have to define this again in `reduced.c`.

mcastelino commented:
Can we split this into two sets of patches. Keeping the generic ACPI changes separate from the hardware reduced ACPI patches. May make it easier to review upstream too. Also we could possibly expand some of the commit messages with more context.

sameo commented:
@bonzini 
> Why are the hardware-reduced tables included, since they are only used by the virt machine type?

It's shows why we are doing the ACPI changes. Would you prefer to have the hardware-reduced tables sent as a standalone, separate patch?

Commit:
hw: acpi: Export the RSDP build API

The hardware-reduced API will need to build RSDP as well, so we should
export this routine.

Cc: "Michael S. Tsirkin" <mst@redhat.com>
Cc: Igor Mammedov <imammedo@redhat.com>
Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Richard Henderson <rth@twiddle.net>
Cc: Eduardo Habkost <ehabkost@redhat.com>
Cc: Marcel Apfelbaum <marcel.apfelbaum@gmail.com>
Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

sameo commented:
All comments addressed. I also added all the relevant maintainers to the commit messages through Cc.
As far as I'm concerned, this is ready to be sent upstream. @rbradford @mcastelino @sboeuf @liujing2 @yangzhon ?

sameo commented:
Patches sent upstream for review:
https://lists.gnu.org/archive/html/qemu-devel/2018-10/msg04953.html

sameo commented:
I reduced the scope of that initial PR and removed the hardware-reduced implementation from it.
Our next patchset will include the hardware-reduced implementation and the virt/arm conversion to it, as requested by @bonzini .

See https://lists.gnu.org/archive/html/qemu-devel/2018-10/msg06209.html

sameo commented:
> I reduced the scope of that initial PR and removed the hardware-reduced implementation from it.
> Our next patchset will include the hardware-reduced implementation and the virt/arm conversion to it, as requested by @bonzini .
> 
> See https://lists.gnu.org/archive/html/qemu-devel/2018-10/msg06209.html

v3 with the right maintainers cc'ed:

https://lists.gnu.org/archive/html/qemu-devel/2018-10/msg06240.html

sameo commented:
v4: https://lists.gnu.org/archive/html/qemu-devel/2018-11/msg00055.html

```
v3 -> v4:
   * Renamed all AcpiConfiguration pointers from conf to acpi_conf.
   * Removed the ACPI_BUILD_ALIGN_SIZE export.
   * Temporarily updated the arm virt build_rsdp() prototype for
     bisectability purposes.
   * Removed unneeded pci headers from acpi-build.c.
   * Refactor the acpi PCI host getter so that it truly is architecture
     agnostic, by carrying the PCI host pointer through the
     AcpiConfiguration structure.
   * Splitted the PCI host AML builder API export patch from the PCI host
     and holes getter one.
   * Reduced the build_srat() export scope to hw/i386 instead of the broader
     hw/acpi. SRAT builders are truly architecture specific and can hardly be
     generalized.
   * Completed the ACPI builder documentation.
```

sameo commented:
v5: https://lists.gnu.org/archive/html/qemu-devel/2018-11/msg00500.html

```
v4 -> v5:
   * Reorganize the ACPI RSDP export and XSDT implementation into 3
     patches.
   * Fix the hw/i386/acpi header inclusions.
```

elmarco committed:
hw/i386: add pc-i440fx-3.1 & pc-q35-3.1

We have a couple of PC_COMPAT_3_0, so we should have 3.1 PC machines,
and update the 3.0 machines to make use of those.

Fixes a "Known issue" from https://wiki.qemu.org/Planning/3.1.

Signed-off-by: Marc-André Lureau <marcandre.lureau@redhat.com>
Reviewed-by: Michael S. Tsirkin <mst@redhat.com>
Reviewed-by: Eduardo Habkost <ehabkost@redhat.com>
Reviewed-by: Igor Mammedov <imammedo@redhat.com>
Message-Id: <20181120132604.22854-1-marcandre.lureau@redhat.com>
Signed-off-by: Eduardo Habkost <ehabkost@redhat.com>

pm215 committed:
Merge remote-tracking branch 'remotes/ehabkost/tags/x86-for-3.1-pull-request' into staging

pc-*-3.1 machine-types

# gpg: Signature made Tue 20 Nov 2018 13:43:24 GMT
# gpg:                using RSA key 2807936F984DC5A6
# gpg: Good signature from "Eduardo Habkost <ehabkost@redhat.com>"
# Primary key fingerprint: 5A32 2FD5 ABC4 D3DB ACCF  D1AA 2807 936F 984D C5A6

* remotes/ehabkost/tags/x86-for-3.1-pull-request:
  hw/i386: add pc-i440fx-3.1 & pc-q35-3.1

Signed-off-by: Peter Maydell <peter.maydell@linaro.org>

rbradford commented:
This project is no longer active and being archived. Cloud Hypervisor is the successor: https://github.com/cloud-hypervisor/cloud-hypervisor
